Manipulator Subsystem

1000DollarBill opened the pull request:
This is all of the mk1 code for the manipulator subsystem, and it is ready for testing once a manipulator prototype is ready.

Commit:
Manipulator Subsystem

Added Manipulator subsystem and made changes to account for it in Robot and RobotMap

Commit:
Added Cube Pickup function

This function is very tentative; there's probably a better way to do it but we'd need to experiment with the manipulator prototype once we have it.

Commit:
Added CubeEject command

Ejects the cube from the manipulator.  Will need work pending experimentation.

Commit:
Added some functions

Set motors to intake, set motors to discharge, read voltage value from the ir sensor.

Commit:
Fixed compilation issue

Import was weird

Commit:
Added conventions

Changed names to match naming conventions and added Javadocs for the functions in Manipulator

strimper13 committed:
Added DropCube and Close Manipulator Commands and Their Functions

DropCube Command and openManipulator function added
CloseManipulator Command and closeManipulator function added

strimper13 committed:
Changed manipulatorSolenoid to a double solenoid

*see above*

Commit:
Merge remote-tracking branch 'xtron202/Manipulator' into Manipulator

Commit:
Renamed a thing

Commit:
Removed unused import

1000DollarBill committed:
Changed type type

manipulatorSolenoidA is now MANIPULATOR_SOLENOID_A etc.

Commit:
Merge remote-tracking branch '1000DollarBill/Manipulator' into Manipulator

1000DollarBill committed:
Fixed dumb

1000DollarBill committed:
Merge remote-tracking branch 'chopshop-166/master' into Manipulator

Commit:
Fixed dumb 2

strimper13 committed:
Merge remote-tracking branch 'chopshop-166/master' into Manipulator

msoucy commented on src/main/java/frc/team166/robot/RobotMap.java:
```diff
@@ -32,11 +34,24 @@
     public static class AnalogInputs {
         //changes input ports into integers
         public final static int tempestgyro = 1;
+        public final static int ir = 2;
```
Use proper casing for name

bot190 replied:
I suggest IR_PORT

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+
+	public void openManipulator() {
+		manipulatorSolenoid.set(Value.kForward);
+		;
```
So wasteful.

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+	 * Sets the motors to 4/5 power inward to take in a cube on the field
+	 */
+	public void setMotorsToIntake() {
+		rollers.set(0.8); //change once you find optimal motor speed
```
Store as a constant at the top of the subsystem

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+
+	//COMMANDS
+	public void initDefaultCommand() {
+		setDefaultCommand(new SubsystemCommand(this) {
```
If the default command does nothing, you can just not call setDefaultCommand

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	public Command DropCube() {
+		return new SubsystemCommand(this) {
```
We didn't talk about this in training, because it's a new thing, but try using an ActionCommand:

`return new ActionCommand("Drop Cube", this, this::openManipulator);`

It's just a tad more concise, and meant specifically for situations when there's one action that's supposed to be run during initialization, then have the command finish

bot190 replied:
Do this for CloseManipulator() below.

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	public Command CubePickup() {
+		return new SubsystemCommand(this) {
```
Use `new SubsystemCommand("Pick Up Cube", this)` instead of just `this`, to give it a name that shows up nicely on smart dashboard

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+import edu.wpi.first.wpilibj.AnalogInput;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.Solenoid;
```
Remove the extra imports, namely this one.

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+
+/**
+ * The 'manipulatorSolenoid' causes the manipulator to open when it's set to 'false' and
+ * the manipulator closes when the 'manipulatorSolenoid' is set to 'true' 
```
This comment seems old and out of date. Also I was hoping for an explanation of the physical behavior so someone could understand how the subsystem works.

Anyway, fix this comment / remove it since the behavior is more standard now.

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+			@Override
+			protected void execute() {
+				//Figure out how to make the motors go 542.865 rpm or 6.81 ft/s without making them spool up
+				if (getAvgEncoderRate() < OPTIMAL_MOTOR_RATE)
```
We should probably be using a PIDController instead of rolling our own.

That said, I think we should push back on the use of encoders here at all. I'm failing to see the benefit given the added complexity.

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,246 @@
+		});
+	}
+
+	public Command DropCube() {
```
Commands should also have Javadocs briefly explaining the operation, and an extended explanation of the steps the command takes.

bot190 reviewed (changes requested):
Replace tabs with spaces.

1000DollarBill committed:
Changes So Deal With IT

ooooooh

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,244 @@
+
+            @Override
+            protected boolean isFinished() {
+                return getAvgEncoderDistance() > 3.0;
```
Make this a preference

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,244 @@
+
+            @Override
+            protected boolean isFinished() {
+                return irSensor.getValue() == 2.0; //Change this once Mech has a prototype
```
Make this a preference

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -25,222 +26,219 @@
+
+        leftIntakeEncoder.setReverseDirection(true);
+
+        motorSpeed = Preferences.getInstance().getDouble(RobotMap.Preferences.MANIPULATOR_MOTOR_INTAKE_SPEED, 0.8);
```
You don't need this `motorSpeed` variable

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -25,222 +26,219 @@
+
+            @Override
+            protected boolean isFinished() {
+                return false;
```
This command never finishes?

bot190 replied:
These should be action commands instead.

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -25,222 +26,219 @@
+
+            @Override
+            protected boolean isFinished() {
+                return false;
```
This command never finishes?

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -25,222 +26,219 @@
+
+            @Override
+            protected void execute() {
+                rollers.set(-0.8); //Change this once you figure out the optimal speed of the motors
```
Use a preference for this

1000DollarBill committed:
*Hits Staples Button* "That Was Easy"

see above^

1000DollarBill committed:
So Gud

berry nice

1000DollarBill committed:
isFinished's fixed for intaking and ejecting the cube

yeet

1000DollarBill committed:
DAB

WoW

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+
+        motorSpeed = Preferences.getInstance().getDouble(RobotMap.Preferences.MANIPULATOR_MOTOR_INTAKE_SPEED, 0.8);
+        motorSpeed = Preferences.getInstance().getDouble(RobotMap.Preferences.MANIPULATOR_MOTOR_DISCHARGE_SPEED, -0.8);
+        cubePickupDistance = Preferences.getInstance().getDouble(RobotMap.Preferences.CUBE_PICKUP_DISTANCE, 0.5);
```
Is `cubePickupDistance` supposed to be the distance to pick up at, or the current distance? You set it above to `getIrDistance()`, but you use it as the preference here

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+
+            @Override
+            protected boolean isFinished() {
+                if (cubePickupDistance < Preferences.getInstance().getDouble(RobotMap.Preferences.CUBE_PICKUP_DISTANCE,
```
Should this be `getIrDistance()`?

1000DollarBill committed:
sause

saucd

msoucy commented:
LGTM :shipit:

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+     * 
+     * @return The voltage from the IR sensor
+     */
+    public double getIRDistance() {
```
This and the four functions below should be made private

msoucy commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+        return new ActionCommand("Drop Cube", this, this::openManipulator);
+    }
+
+    public Command CloseManipulator() {
```
Since we already know that it's part of the manipulator subsystem, this can just be called `Close`.

1000DollarBill replied:
I disagree, we have always wanted our code to be as clear as possible, and it is inane to delete 11 characters for "cleaner code."

bot190 commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -24,10 +25,11 @@
+	//this just adds a drive subsystem
+	public static final Drive drive = new Drive();
+	public static final Manipulator manip = new Manipulator();
+	public static OI m_oi;
```
These are still tabs not spaces..

bot190 commented on src/main/java/frc/team166/robot/RobotMap.java:
```diff
@@ -32,11 +34,29 @@
+    public static class Encoders {
+        public static final int LEFT_ROLLER_B = 6;
+        public static final int RIGHT_ROLLER_B = 7;
+    }
```
Get rid of these constants now that we don't have encoders on the rollers

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+        addChild(irSensor);
+
+        motorSpeed = Preferences.getInstance().getDouble(RobotMap.Preferences.MANIPULATOR_MOTOR_INTAKE_SPEED, 0.8);
+        motorSpeed = Preferences.getInstance().getDouble(RobotMap.Preferences.MANIPULATOR_MOTOR_DISCHARGE_SPEED, -0.8);
```
Do you need these? You just get the preference value directly later on.

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+            @Override
+            protected void execute() {
+
+            }
```
This can just be:
protected void execute() {}

msoucy replied:
Not with the autoformatting - it CAN be removed, however

bot190 commented on src/main/java/frc/team166/robot/subsystems/Manipulator.java:
```diff
@@ -0,0 +1,168 @@
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.Preferences;
+import edu.wpi.first.wpilibj.Solenoid;
```
Remove this unused import.

bot190 reviewed (changes requested):
Some more comments.

1000DollarBill committed:
Fixed Manipulator Crap Again, Please GOD Let This Be The Last Pull Request For This Subsystem

strimper13 committed:
Fixed Matt's Issues (Review Comments)

Edited four public functions to be private instead (Starting at line 55). Also Took out four lines of useless garbage about unused preferences.

1000DollarBill committed:
Fixed Bulls**t

see above^

1000DollarBill committed:
Preferences Put Back Into Constructer

bot190 committed:
Merge branch 'master' into Manipulator
